### PR TITLE
Refactor the JAX dispatcher

### PR DESCRIPTION
**Pull request by rlouf**
In this PR we refactor the JAX dispatcher to mirror the way files are organized in the Numba dispatcher.

Here are a few important guidelines and requirements to check before your PR can be merged:
+ [x] There is an informative high-level description of the changes.
+ [NA] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [x] [`pre-commit`](https://pre-commit.com/#installation) is installed and [set up](https://pre-commit.com/#3-install-the-git-hook-scripts).
+ [x] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [x] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), and there are **no commits that fix changes introduced by other commits in the same branch/BR**.
+ [x] There are tests covering the changes introduced in the PR.

### CODE REVIEW

**Comment by rlouf** [comment]
This is ready for review, Mypy complains that `Module "numpy.random.bit_generator" has no attribute "_coerce_to_uint32_array"  [attr-defined]` which does not make sense to me (especially since it does not complain on `main`).

**Comment by brandonwillard** [comment]
> This is ready for review, Mypy complains that `Module "numpy.random.bit_generator" has no attribute "_coerce_to_uint32_array" [attr-defined]` which does not make sense to me (especially since it does not complain on `main`).

You can add an ignore for that.

File: aesara/link/jax/dispatch/basic.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+from aesara.tensor.shape import Reshape, Shape, Shape_i, SpecifyShape, Unbroadcast
+from aesara.tensor.subtensor import (
+    AdvancedIncSubtensor,
+    AdvancedIncSubtensor1,
+    AdvancedSubtensor,
+    AdvancedSubtensor1,
+    IncSubtensor,
+    Subtensor,
+    indices_from_subtensor,
+)
```
Comment (brandonwillard) [inline aesara/link/jax/dispatch/basic.py:24]: Do you want to make `dispatch` modules for these in a follow-up PR?  Regardless, we should create an issue for these.

Comment (rlouf) [reply aesara/link/jax/dispatch/basic.py:24]: Do you mean `jax/dispatch/subtensor.py` and `jax/dispatch/shape.py`? I'm happy to do it in this PR.

Comment (brandonwillard) [reply aesara/link/jax/dispatch/basic.py:24]: > Do you mean `jax/dispatch/subtensor.py` and `jax/dispatch/shape.py`? I'm happy to do it in this PR.

Yes, those two.

Comment (rlouf) [reply aesara/link/jax/dispatch/basic.py]: I just pushed the changes.

**Review by brandonwillard — APPROVED** [review]

**Pull request merged**